### PR TITLE
[all] PSL branch prerequisites

### PR DESCRIPTION
**Pull request by bcarrez**
To review after #471 is merged (DONE) You can now review it :) 

This PR includes all changes and adds of the incoming [PSL] branch that can be merged directly. It includes some Python bindings, additional tests, and various cleanings.
Hopefully, it should not break anything.

 in BaseContext add variants of getObjects() which ease the syntax (eg: for(auto& i : getNodeObjects()){...} + corresponding tests. 

- in Node add variants of getNodeObjects() and getTreeObjects that does not force to ease the syntax+  corresponding tests. 

- in BaseLink add a setFlag method and a setPersistant...this makes things more consistent with BaseData.

- in Binding_Base add the methods: addNewData, addData, getData, getLink, getListOfDataFields, getListOfLinks. The function are documented with docstring and tested in test_BindingBase.py

- in Binding_Data add methods: isSet, isPersistant, setPersistant, hasParent, getParentPath, setValueString. The function are documented with docstring and tested in test_BindingData.py
 
- in Binding_Link add methods: setValueString, isPersistant, setPersistant. The function are documented with docstring and tested in test_BindingLink.py. 

- in Binding_SofaModule add methods: getAliasesFor. The function are documented with docstrings & tested in test_BindingSofa.py

- in SofaPython add functions: sendMessageFromException, getSofaFormattedStringFromException

- in SofaTest/_init_.py add ASSERT_* functions that generates real gtest failure. 

- in runSofa in QDataSimpleEdit & QLinkSimpleEdit correctly prevent writing when the data field is readOnly
______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by bcarrez** [comment]
The problem seems to be generated by the merge with master. I will fix this merge and push a fixed version.

**Comment by damienmarchal** [comment]
OMG There is so much things to do to finalize this PR. 
I'm reviewing my own code but among things:
- add docstring,
- remove the todo before PR
- add tests to each new features.
- ...

**Comment by damienmarchal** [comment]
Ouch this PR is getting old. 
@bcarrez If you have time today it would be nice to finalize this PR so that we can move forward :)

**Comment by damienmarchal** [comment]
Ouch I merged upstream/master & #471 so the diff is totally unreadable. Hope it will get better 
when 471 is merged.

**Comment by damienmarchal** [comment]
Hey @bcarrez I fixed all the things I wanted to do. We should move forward now.

**Comment by damienmarchal** [comment]
Hi everyone, this this is an "all in one" branch that adds functions in Sofa 
before we can make a PR with the PSL plugin.

The PR is two month old...no one complains and now all the added function are tested & documented so it may be merged soon.

**Comment by guparan** [comment]
Are you done with the rebase?

**Comment by damienmarchal** [comment]
I'm waiting the CI to finish but I think we are all done now.

**Comment by damienmarchal** [comment]
@guparan, @fredroy, @bcarrez , @hugtalbot 
Hello, this branch is now synced with master, CI is ok, added functions have theirs tests...
So it is waiting.

**Comment by damienmarchal** [comment]
No answer so "qui tacet consentire videtur" :)

File: applications/plugins/SofaPython/Binding.cpp [hunk 1]
```diff
@@ -111,15 +110,16 @@ void bindSofaPythonModule()
     SP_ADD_CLASS_IN_FACTORY(PointSetTopologyModifier,sofa::component::topology::PointSetTopologyModifier)
     SP_ADD_CLASS_IN_FACTORY(TriangleSetTopologyModifier,sofa::component::topology::TriangleSetTopologyModifier)
 
-    /// Custom Exception to embed
+    /// Custom Exception
+    const char* name = "Sofa.SofaException";
```
Comment (jnbrunet) [inline applications/plugins/SofaPython/Binding.cpp]: Were those variables (name and doc) supposed to be use somewhere?

**Review by jnbrunet — APPROVED** [review]

**Comment by damienmarchal** [comment]
So I merge :)

**Comment by guparan** [comment]
Houston we got a problem: https://ci.inria.fr/sofa-ci/job/ubuntu_clang-3.8_options-debug/545/testReport/

**Pull request merged**